### PR TITLE
[AMDGPU] Add tests for SIPreAllocateWWMRegs

### PR DESCRIPTION
**Pull request by optimisan**

### CODE REVIEW

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[DEF:%[0-9]+]]:vgpr_32 = IMPLICIT_DEF
+    ; CHECK-NEXT: renamable $sgpr4_sgpr5 = ENTER_STRICT_WWM -1, implicit-def $exec, implicit-def $scc, implicit $exec
+    ; CHECK-NEXT: $vgpr0 = V_MOV_B32_e32 0, implicit $exec
+    ; CHECK-NEXT: dead $vgpr0 = V_MOV_B32_dpp $vgpr0, [[DEF]], 323, 12, 15, 0, implicit $exec
+    ; CHECK-NEXT: $exec = EXIT_STRICT_WWM killed renamable $sgpr4_sgpr5
+    ; CHECK-NEXT: dead [[COPY:%[0-9]+]]:vgpr_32 = COPY [[DEF]]
+    %0:vgpr_32 = IMPLICIT_DEF
+    renamable $sgpr4_sgpr5 = ENTER_STRICT_WWM -1, implicit-def $exec, implicit-def $scc, implicit $exec
+    %24:vgpr_32 = V_MOV_B32_e32 0, implicit $exec
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:22]: Use run-pass=none to compact the register numbers

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn -verify-machineinstrs -run-pass=si-pre-allocate-wwm-regs -o - -mcpu=tahiti %s  | FileCheck %s
+
+---
+
+name: pre_allocate_wwm_regs_strict
+tracksRegLiveness: true
+body: |
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:8]: This really needs checks for the MFI serialized allocated register. It is not yet serialized, but it really needs to be (and is currently causing issues for me)

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-sgpr-spills.mir [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  bb.0:
+    liveins: $sgpr1
+    ; CHECK-LABEL: name: pre_allocate_wwm_spill_to_vgpr
+    ; CHECK: liveins: $sgpr1
+    ; CHECK-NEXT: {{  $}}
+    ; CHECK-NEXT: [[DEF:%[0-9]+]]:vgpr_32 = IMPLICIT_DEF
+    ; CHECK-NEXT: dead $vgpr0 = SI_SPILL_S32_TO_VGPR $sgpr1, 0, [[DEF]]
+    ; CHECK-NEXT: dead [[COPY:%[0-9]+]]:vgpr_32 = COPY [[DEF]]
+    %0:vgpr_32 = IMPLICIT_DEF
+    %23:vgpr_32 = SI_SPILL_S32_TO_VGPR $sgpr1, 0, %0:vgpr_32
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-sgpr-spills.mir:18]: Compact register numbers

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-sgpr-spills.mir [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn -verify-machineinstrs -amdgpu-prealloc-sgpr-spill-vgprs -run-pass=si-pre-allocate-wwm-regs -o - -mcpu=tahiti %s | FileCheck %s
+
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-sgpr-spills.mir:3]: Why does this need to be split into a separate test file?

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir (hunk 2 above)
Comment (Akshat-Oke) [reply llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:8]: What exactly should be serialized for allocated registers?
I am working on serializing virtual register flags (currently there is only one WWM_REG)

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:8]: WWMSpills, SpillVGPRs, SpillPhysVGPRs, SGPRSpillsTo*

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    liveins: $sgpr1
+    ; CHECK2-LABEL: name: pre_allocate_wwm_spill_to_vgpr
+    ; CHECK2: liveins: $sgpr1
+    ; CHECK2-NEXT: {{  $}}
+    ; CHECK2-NEXT: [[DEF:%[0-9]+]]:vgpr_32 = IMPLICIT_DEF
+    ; CHECK2-NEXT: dead $vgpr0 = SI_SPILL_S32_TO_VGPR $sgpr1, 0, [[DEF]]
+    ; CHECK2-NEXT: dead [[COPY:%[0-9]+]]:vgpr_32 = COPY [[DEF]]
+    %0:vgpr_32 = IMPLICIT_DEF
+    %1:vgpr_32 = SI_SPILL_S32_TO_VGPR $sgpr1, 0, %0
+    %2:vgpr_32 = COPY %0
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:43]: Missing ... at end of function

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
+# RUN: llc -mtriple=amdgcn -verify-machineinstrs -run-pass=si-pre-allocate-wwm-regs -o - -mcpu=tahiti %s  | FileCheck %s
+# RUN: llc -mtriple=amdgcn -verify-machineinstrs -amdgpu-prealloc-sgpr-spill-vgprs -run-pass=si-pre-allocate-wwm-regs -o - -mcpu=tahiti %s | FileCheck %s --check-prefix=CHECK2
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:3]: Move the -mcpu together with -mtriple

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+---
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:5]: Add fixme to check the MachineFunctionInfo reserved register information

**Review by arsenm — APPROVED** [review]

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir (hunk 7 above)
Comment (optimisan) [reply llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:5]: Is it for WWM reserved registers?

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:5]: Yes

Comment (optimisan) [reply llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:5]: But we have that already(?)
```
wwmReservedRegs:
  - '$vgpr0'
```

Comment (arsenm) [reply llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:5]: Then you need to write manual checks for it. Update_mir_test_checks doesn't currently support the function level properties

Comment (optimisan) [reply llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:6]: Added a comment for the manual check.

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir [hunk 8]
```diff
@@ -0,0 +1,49 @@
+# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
```
Comment (arsenm) [inline llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:1]: Best to remote this and remove the comment then

**Comment by optimisan** [comment]
### Merge activity

* **Oct 22, 5:41 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/109963).

File: llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir (hunk 8 above)
Comment (optimisan) [reply llvm/test/CodeGen/AMDGPU/si-pre-allocate-wwm-regs.mir:1]: What is remote?

**Pull request merged**